### ISSUE TITLE
Bug/question: E2E testing with GraphQL gives a 401 error

### ISSUE THREAD

**Issue by Ulset** [Shopify/shopify-app-js#1756]
Hi we are working on a shopify app, and as our codebase grows the need for testing E2E rises.

Im currently trying to implement some very basic testing of our backend queries to Shopify. Im initalizing the shopify wrapper using the `shopifyApp` function and everything runs smoothly. After initalizing the wrapper with test config:
```javascript
const prismaSessionStorage = new PrismaSessionStorage(prisma);

let config: Parameters<typeof shopifyApp>[0] | null

if (process.env.NODE_ENV === 'test') {
  config = testConfig({
    sessionStorage: prismaSessionStorage
  });
} else {
   // Actual config
}
const shopify_handler = shopifyApp(config);
```

The following test runs smoothly:
```javascript
test("Verify that queries are okay", async () => {
  const someSession = await prisma.session.findFirstOrThrow()
  const {admin} = await shopify_handler.unauthenticated.admin(someSession.shop)
  const shop = await admin.rest.get({path: 'shop.json'}).then((r) => r.json())
  console.log(shop)
})
```
I can see in my logs my current dev store
![Image](https://github.com/user-attachments/assets/6a4e03cb-b5be-4f01-934d-af19fcdc6476)

The problem seems to arise when i try to fetch anything from the graphql API

Given the following modified code
```javascript
test("Verify that queries are okay", async () => {
  const someSession = await prisma.session.findFirstOrThrow()
  const {admin} = await shopify_handler.unauthenticated.admin(someSession.shop)
  const shop = await admin.graphql(`
    #graphql
    query {
      shop {
        currencyCode
        myshopifyDomain
        shopOwnerName
        email
        contactEmail
        plan {
          partnerDevelopment
        }
      }
    }
  `).then((r) => r.json()).then((data) => {
    return data.data.shop
  }).catch((e) => {
    console.error('Error fetching shop info', e)
    return null
  })
  console.log(shop)
})
```
I seem to get a 401 unauthenticated error:
![Image](https://github.com/user-attachments/assets/bcd8958e-d00b-4f22-82f4-a7ffbfb20ff6)

Anyone have a solution for this?

**Comment by sle-c** [comment]
Hi @Ulset ,

I tried to replicate the test with the following code

```javascript
import "dotenv/config";
import { test } from "@jest/globals";
import prisma from "../app/db.server";
import { unauthenticated } from "../app/shopify.server";

test("Verify that queries are okay", async () => {
  const someSession = await prisma.session.findFirstOrThrow();
  const { admin } = await unauthenticated.admin(someSession.shop);
  const shop = await admin.rest.get({ path: "shop.json" }).then((r) => r.json());
  console.log(shop);
});

test("Verify that gql queries are okay", async () => {
  const someSession = await prisma.session.findFirstOrThrow();
  const { admin } = await unauthenticated.admin(someSession.shop);
  const shop = await admin.graphql(`
    #graphql
    query {
      shop {
        currencyCode
        myshopifyDomain
        shopOwnerName
        email
        contactEmail
        plan {
          partnerDevelopment
        }
      }
    }
  `).then((r) => r.json()).then((data) => {
    return data.data.shop
  }).catch((e) => {
    console.error('Error fetching shop info', e)
    return null
  })
  console.log(shop)
})
```

When I ran the test, both test cases passed. I used the default shopifyApp config from the template's `shopify.server` file. I didn't have a separate test configuration. Can you check if the sessions are still valid for the test config you used?

Si

**Comment by sle-c** [comment]
Hi @Ulset ,

@lizkenyon and I looked into this issue deeper and we found an inconsistent behaviour between the REST client and GraphQL Client. Thank you for reporting this issue and the `testConfig` detail!

TLDR of the issue:
- In the REST client, we have logic to use the real session token from the passed in session. Although it also check for another access token in the config before using the session access token.
- In the GraphQL client, we do not have that logic and it always use the other access token in the config which leads to a `401`.

Next steps:
- We'll put in a fix to make the 2 clients behaves the same way for consistency.
- Ideally, when using `testConfig`, one should **not** be able to make a request to a real production shop. It's intended for unit testing. The requests should be mocked. 
  - For E2E tests, one should use a real config to a production shop, not a test config.

**Comment by byrichardpowell** [comment]
I'm going to close this issue as there hasn't been any activity for a while
